### PR TITLE
Implement PlayerList

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Implemented as far as possible

### CODE REVIEW

File: src/game/common/rts/playerlist.cpp [hunk 1]
```diff
@@ -18,3 +18,234 @@
...
+    for (auto i = 0; i < MAX_PLAYER_COUNT; ++i) {
+        m_players[i] = new Player(i);
+    }
+    Init();
+}
+
+// zh: 0x0045A8D0 wb: 0x007BEEFD
+PlayerList::~PlayerList()
+{
+    Init();
```
Comment (tomsons26) [inline src/game/common/rts/playerlist.cpp]: Add a todo to investigate as this is quite illogical

File: src/game/common/rts/playerlist.cpp [hunk 2]
```diff
@@ -18,3 +18,234 @@
...
+    for (auto i = 0; i < Get_Player_Count(); ++i) {
+        auto *player = Get_Nth_Player(i);
+        if (player == nullptr || player == target_player) {
+            continue;
+        }
+
+        const auto relationship = target_player->Get_Relationship(player->Get_Default_Team());
+        switch (relationship) {
+            case Relationship::ENEMIES:
+                if (relationship_mask & (1ULL << 2)) {
```
Comment (tomsons26) [inline src/game/common/rts/playerlist.cpp]: why is this more preferable than just & 4?

Comment (duncanspumpkin) [reply src/game/common/rts/playerlist.cpp]: These are flags. I was going to switch them to an enum but forgot. Will add that shortly.

File: src/game/common/rts/playerlist.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/game/common/rts/playerlist.cpp]: Init clears out a lot of things and unloads the PlayerTemplate I think that is the reason for it. Init seems to get called a lot.

File: src/game/common/rts/playerlist.h [hunk 3]
```diff
@@ -18,46 +18,71 @@
 #include "snapshot.h"
 #include "subsysteminterface.h"
 
+class Team;
+
+constexpr int32_t MAX_PLAYER_COUNT = 16;
```
Comment (tomsons26) [inline src/game/common/rts/playerlist.h]: should be a in a more global header, its gonna be used a lot

Comment (duncanspumpkin) [reply src/game/common/rts/playerlist.h]: Any suggestions as to which one?

Comment (duncanspumpkin) [reply src/game/common/rts/playerlist.h]: Suggested gametypes.h

File: src/game/common/rts/playerlist.h [hunk 4]
```diff
@@ -48,7 +56,8 @@ class PlayerList : public SubsystemInterface, public SnapShot
     void Set_Local_Player(Player *player);
     Player *Get_Player_From_Mask(uint16_t mask);
     Player *Get_Each_Player_From_Mask(uint16_t &mask);
-    uint16_t Get_Players_With_Relationship(int32_t player_idx, uint16_t relationship_mask);
+    uint16_t Get_Players_With_Relationship(
+        int32_t player_idx, uint16_t relationship_mask); // Uses PlayerRelationshipFlags for mask
```
Comment (xezon) [inline src/game/common/rts/playerlist.h]: You could get rid of the comment by creating a `using relationship_flags_t = uint16_t` or similar nearby the flag definition.

File: src/game/common/rts/player.h [hunk 5]
```diff
@@ -18,22 +18,60 @@
...
+    NameKeyType m_playerNameKey;
     int m_playerIndex;
-    char unk[0xFC];
+    char unk4[0xFC];
     int m_playerColor;
-};
\ No newline at end of file
+    uint8_t unk5[0x160 - 0x128];
+    Team *m_defaultTeam;
+    uint8_t unk6[0x44C - 0x164]; // There is a size 4 field specific to WB in here
```
Comment (xezon) [inline src/game/common/rts/player.h]: WorldBuilder specific bytes should not be compiled into non-`GAME_DEBUG_STRUCTS` afaik.

```c++
#ifdef GAME_DEBUG_STRUCTS
uint8_t unk6[0x44C - 0x164];
#else
uint8_t unk6[0x44C - 0x164 - 4];
#endif
```

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**